### PR TITLE
Add `Monoid` and `Semigroup` instances

### LINKED ISSUE #269: Monoid instance
It seems to me that `Graph` is missing a Monoid instance:

```haskell
mempty = empty
(<>) = overlay
```

There's also a monoid over `connect`, but that doesn't seem to be as relevant in terms of how people think about constructing graphs --- which is adding the vertices and edges to a bucket.

### PR DESCRIPTION
**Pull request by ocharles**
Fixes #269.

### CODE REVIEW

File: src/Algebra/Graph.hs [hunk 1]
```diff
@@ -294,6 +294,12 @@ instance MonadPlus Graph where
     mzero = Empty
     mplus = Overlay
 
+instance Semigroup (Graph a) where
```
Comment (snowleopard) [inline src/Algebra/Graph.hs:297]: Let's add a clarification (here and in other instances too):

```suggestion
-- | Defined via 'overlay'.
instance Semigroup (Graph a) where
```

**Comment by snowleopard** [comment]
@ocharles Many thanks for the PR! Alas, CI fails on a couple of GHC versions, most likely we need to do a dance with imports and/or CPP.

**Comment by ocharles** [comment]
Do you really care about 8.0 and 8.2? You might be inventing more work than you need to, that's quite the range of support! But that's for another issue, I'll get this all building with CI tomorrow (along with those instance comments) :+1:

**Comment by snowleopard** [comment]
Yeah, I think you are right and I should just stop supporting 8.0 and 8.2.

Shall we keep this PR as is for now and switch to dropping 8.0 and 8.2 to avoid unnecessary work? Either way is fine for me.

**Comment by ocharles** [comment]
@snowleopard If you're planning to drop support for 8 & 8.2 then I won't make any changes - I'm always one for doing the least amount of work possible! :laughing: I will add the Haddock comments though.

**Comment by snowleopard** [comment]
@ocharles Done!

**Comment by ocharles** [comment]
@snowleopard ok, documentation added

File: src/Algebra/Graph/Labelled.hs [hunk 2]
```diff
@@ -92,6 +92,14 @@ instance (NFData e, NFData a) => NFData (Graph e a) where
     rnf (Vertex  x    ) = rnf x
     rnf (Connect e x y) = e `seq` rnf x `seq` rnf y
 
+-- | Defined via 'overlay'.
+instance Monoid e => Semigroup (Graph e a) where
+    (<>) = overlay
+
+-- | Defined via 'empty'.
```
Comment (snowleopard) [inline src/Algebra/Graph/Labelled.hs:99]:
```suggestion
-- | Defined via 'overlay' and 'empty'.
```

**Comment by snowleopard** [comment]
@ocharles Thanks, merged! Just tweaked one of the comments.

**Pull request merged**